### ISSUE TITLE
Display exact match search result at the top and with clear label

### ISSUE THREAD

**Issue by joelverhagen** [NuGet/NuGetGallery#7463]
Thanks @zkat for the hint in the demo day a while back.

We have deprioritized exact match in some cases but we could bring it back with some UI tweaks like npm:

![a](https://user-images.githubusercontent.com/94054/63299806-8048ad00-c28b-11e9-9311-a871f0078c4d.png)

We would need to think through this analogous UI on nuget.org and on VS client. Might need a protocol enhancement.

**Comment by zkat** [comment]
## crates.io

I particularly like the way crates.io does it, because they make it clear something is top-of-the-list because it's an -exact match-, not because it ranked better. It's common to want that exact match, after all.

![{7D437DB1-1A2C-4855-80DE-86BD718D52C8}](https://user-images.githubusercontent.com/17535/63300109-488e3500-c28c-11e9-9a4a-2e35399cefce.png)

## rubygems.org

rubygems.org does this too, but I'm less of a fan of displaying it twice:

![{B431FC9A-A8DE-4CEE-B685-9173CF5933F9}](https://user-images.githubusercontent.com/17535/63300256-9b67ec80-c28c-11e9-9c01-5664ad17ea7e.png)

**Comment by 304NotModified** [comment]
I like this idea! I prefer the label (npm style)

**Comment by loic-sharma** [comment]
Here is another example that would be helped by this work item: https://github.com/NuGet/NuGetGallery/issues/7518

**Comment by mv10-work** [comment]
Not listing exact matches first is completely counter-intuitive.

> You really need to prioritize this 'feature' because without it's incredibly hard to find a package that has any part of the name that's common (like Web or AspNet anything).

Or the 500-pound gorilla of popularity-ranking-spam, "Microsoft".

Yesterday I had to search for `Microsoft.AspNetCore.Components.Authorization` which currently produces 58,680 hits, and the exact match doesn't show up until page 7 on NuGet.org ... no idea where it shows up in the VS GUI, I didn't have that kind of patience.

**Comment by joelverhagen** [comment]
I've shipped the first step in this work to our primary region. This means searches on nuget.org and in Visual Studio in non-China should have this first fix.

When a search query has an exact match _and_ the  package ID contains symbols (dot `.`, underscore `_`, hyphen `-`) then that package is at the top.

Regarding package IDs without symbols, we chose not to put that one always at the top yet because:

1. A lot of our top queries have low download count exact match "false positives". Without clear UI saying "this is at the top because exact match only" our telemetry and user feedback indicates that we shouldn't do this one yet.
1. We don't have our A/B testing infrastructure for search changes ready yet so we wanted to avoid changes that we are less confident about.
1. All of the exact match issues reported since we shipped new search were concerning package IDs with dots. In other words, this small fix covers the reported pain points.

To be clear, exact match means that the search query exactly matches a package ID. If the query is "system text json", System.Text.Json isn't _necessarily_ at the top but, in this case, it is for other scoring reasons. If there are two package IDs in the search query or a prefix of a package ID the improvement I made does not apply. Other relevancy changes are planned to improve these non-exact match cases but they are less urgent.

As always, we warmly welcome input on how we are doing. For the exact match case, please feel free to chime in here. For other search relevancy issues, please respond to https://github.com/NuGet/NuGetGallery/issues/4124.

This issue tracks the UI work and putting all exact matches at the top so I will leave it open until we do that. Note the UI is a little challenging because we are thinking about all of the other "markers" or "badges" that may be on a search result so we want to make a UI change that gels well.

**Comment by jariq** [comment]
Sorry I don't have time to read about the problem you are trying to fix, but I wanted to let you know that I find current search results confusing and quite hard to understand. Totally unrelated packages appear higher in the result without any obvious reason. Documented in [Screenshot_2019-09-19 NuGet Gallery Packages matching Pkcs11Interop](https://user-images.githubusercontent.com/1163818/65214189-514f7200-daa9-11e9-9290-8933aab07052.png).

**Comment by normj** [comment]
I can confirm our 2 most problematic packages Amazon.AspNetCore.Identity.Cognito and Amazon.AspNetCore.DataProtection.SSM are showing up at the top of this list now.

**Comment by jariq** [comment]
@karann-msft  judging solely by your reaction to my previous comment I'm guessing I wasn't communicating the problem clear enough. Let me try again.

When I searched for `Pkcs11Interop` before this change, I got `Pkcs11Interop` package as the 1st and related package `Pkcs11Interop.X509Store` as 2nd or at most as 3rd. 

When I search for `Pkcs11Interop` now I got `Pkcs11Interop` package as the 1st (no change here) but related package `Pkcs11Interop.X509Store` is 9th and there are many unrelated packages between the 1st and 9th place.

So in my point of view this change may have fixed exact match but it messed up the rest of the result.

**Comment by 304NotModified** [comment]
PS. There is an comparison page (old/new search results) here: https://www.nuget.org/experiments/search-sxs

**Comment by loic-sharma** [comment]
Hey @jariq, the search service splits inputs like `Pkcs11Interop` into tokens like `Pkcs`, `11`, and `Interop`. All of the "unrelated" results you mention match `Pkcs` or `Interop`, so this behavior is expected. What's happening here is that we're favoring downloads a little too much for this specific query. This is a tricky to balance, but we have a few improvements in the works that should help this query. Stay tuned!

**Comment by loic-sharma** [comment]
This would also help clarify the [`json.net` search](https://www.nuget.org/packages?q=json.net): the first result is the package [`Json.Net`](https://www.nuget.org/packages/Json.Net/) as it is an exact match on the package ID, the second result is the package [`Newtonsoft.Json`](https://www.nuget.org/packages/Newtonsoft.Json/) as its package title is `Json.NET`.

/cc @anangaur

**Comment by etvorun** [comment]
Search result includes "exact match" badge now.
